libraries/plugins/persistent_proposals/persistent_proposals_plugin.cpp

```diff
@@ -0,0 +1,206 @@
+        auto proposal_update_base_it = proposal_update_base_by_original_id.find(proposal.id);
+        if (proposal_update_base_it == proposal_update_base_by_original_id.end())
+        {
+            create_update_base(proposal);
+        }
+
+        _database->create<proposal_update_object>([&](proposal_update_object& update) {
+            update.proposal = proposal.id;
+
+            update.added_owner_approvals = get_difference(proposal.available_owner_approvals, proposal_update_base_it->available_owner_approvals);
+            update.removed_owner_approvals = get_difference(proposal_update_base_it->available_owner_approvals, proposal.available_owner_approvals);
+            update.added_active_approvals = get_difference(proposal.available_active_approvals, proposal_update_base_it->available_active_approvals);
+            update.removed_active_approvals = get_difference(proposal_update_base_it->available_active_approvals, proposal.available_active_approvals);
+            update.added_key_approvals = get_difference(proposal.available_key_approvals, proposal_update_base_it->available_key_approvals);
+            update.removed_key_approvals = get_difference(proposal_update_base_it->available_key_approvals, proposal.available_key_approvals);
+
+            update.update_time = fc::time_point::now();
+        });
+
+        _database->modify(*proposal_update_base_it, [&](proposal_update_base_object& update_base) {
+            update_base.original_id = proposal.id;
+
+            duplicate_proposal(proposal, &update_base);
+        });
+    }
+
+    virtual void on_remove( const object& obj ) override
+    {
+        const auto& proposal = dynamic_cast<const proposal_object&>(obj);
+        if (!is_expired(proposal) && !proposal.is_authorized_to_execute(*_database))
+        {
+            return;
+        }
+
+        auto& persistent_proposal_by_original_id = _database->get_index_type<persistent_proposal_index>().indices().get<by_original_id>();
+
+        auto old_persistent_proposal_it = persistent_proposal_by_original_id.find(proposal.id);
+        if (old_persistent_proposal_it != persistent_proposal_by_original_id.end())
+        {
+            _database->remove(*old_persistent_proposal_it);
+        }
+
+        _database->create<persistent_proposal_object>([&](persistent_proposal_object& persistent_proposal) {
+            persistent_proposal.original_id = proposal.id;
+
+            duplicate_proposal(proposal, &persistent_proposal);
+        });
+    }
+
+private:
+    bool is_expired(const proposal_object& proposal)
+    {
+        return proposal.expiration_time <= _database->head_block_time();
+    }
+
+    template <class T>
+    flat_set<T> get_difference(const flat_set<T>& left, const flat_set<T>& right)
+    {
+        flat_set<T> difference;
+
+        for (const auto& item: left)
+        {
+            if (right.count(item) == 0)
+            {
+                difference.insert(item);
+            }
+        }
```

Comment:
Not really, because this method is called here:
```

            update.added_owner_approvals = get_difference(proposal.available_owner_approvals, proposal_update_base_it->available_owner_approvals);
            update.removed_owner_approvals = get_difference(proposal_update_base_it->available_owner_approvals, proposal.available_owner_approvals);
            update.added_active_approvals = get_difference(proposal.available_active_approvals, proposal_update_base_it->available_active_approvals);
            update.removed_active_approvals = get_difference(proposal_update_base_it->available_active_approvals, proposal.available_active_approvals);
            update.added_key_approvals = get_difference(proposal.available_key_approvals, proposal_update_base_it->available_key_approvals);
            update.removed_key_approvals = get_difference(proposal_update_base_it->available_key_approvals, proposal.available_key_approvals);
```
those lists can only contain things, e.g. for removal of an id, you would "add" and id to the `removed_*_approvals` list.